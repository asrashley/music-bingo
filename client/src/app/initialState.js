import { initialState as cards } from '../cards/cardsSlice';
import { initialState as games } from '../games/gamesSlice';
import { initialState as tickets } from '../tickets/ticketsSlice';
import { initialState as user } from '../user/userSlice';

export const initialState = Object.freeze({
    cards,
    games,
    tickets,
    user,
});
