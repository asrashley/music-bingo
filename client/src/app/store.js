import { configureStore, getDefaultMiddleware } from '@reduxjs/toolkit';
import logger from 'redux-logger';
import { connectRouter } from 'connected-react-router';
import { createBrowserHistory } from 'history';
import { reducer as formReducer } from 'redux-form';

import cardsReducer from '../cards/cardsSlice';
import gamesReducer from '../games/gamesSlice';
import ticketsReducer from '../tickets/ticketsSlice';
import userReducer from '../user/userSlice';
import adminReducer from '../admin/adminSlice';
import usersMiddleware from '../user/userMiddleware';

export const history = createBrowserHistory();

export const store = configureStore({
  middleware: [...getDefaultMiddleware(), usersMiddleware, logger],
  reducer: {
    admin: adminReducer,
    cards: cardsReducer,
    form: formReducer,
    games: gamesReducer,
    tickets: ticketsReducer,
    router: connectRouter(history),
    user: userReducer,
  },
});
