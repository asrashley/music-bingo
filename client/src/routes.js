
export default {
  index: '/',
  login: '/user/login',
  logout: '/user/logout',
  passwordReset: '/user/reset',
  passwordResetConfirm: '/user/reset/:token',
  register: '/register',
  game: '/game/:gamePk',
  play: '/game/:gamePk/tickets',
  viewTicket: '/game/:gamePk/:ticketPk',
  pastGames: '/history',
  trackListing: '/history/:gamePk',
  listUsers: '/users'
};

