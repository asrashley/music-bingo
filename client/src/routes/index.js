export default {
  index: '/',
  login: '/user/login',
  logout: '/user/logout',
  user: '/user',
  changeUser: '/user/modify',
  passwordReset: '/user/reset',
  passwordResetConfirm: '/user/reset/:token',
  guestLinks: '/user/guests',
  listUsers: '/user/users',
  guestAccess: '/invite/:token',
  register: '/register',
  listDirectories: '/directories',
  listGames: '/game',
  chooseTickets: '/game/:gameId',
  play: '/game/:gameId/tickets',
  viewTicket: '/game/:gameId/:ticketPk',
  pastGames: '/history',
  trackListing: '/history/:gameId',
  privacy: '/privacy',
};
