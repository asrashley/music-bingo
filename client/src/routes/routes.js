

export const routes = Object.freeze({
  index: '/',
  login: '/user/login',
  logout: '/user/logout',
  user: '/user',
  changeUser: '/user/modify',
  passwordReset: '/user/reset',
  passwordResetConfirm: '/user/reset/:token',
  guestLinks: '/user/guests',
  settingsIndex: '/user/settings',
  settingsSection: '/user/settings/:section',
  listUsers: '/user/users',
  guestAccess: '/invite/:token',
  register: '/register',
  listDirectories: '/clips',
  listDirectory: '/clips/:dirPk',
  listGames: '/game',
  chooseTickets: '/game/:gameId',
  play: '/game/:gameId/tickets',
  viewTicket: '/game/:gameId/tickets/:ticketPk',
  pastGamesPopularity: '/history',
  pastGamesList: '/history/games',
  pastGamesCalendar: '/history/calendar',
  pastGamesCalendarByYear: '/history/calendar/:year',
  pastGamesByTheme: '/history/themes/:slug',
  gameLastUsed: '/history/themes',
  trackListingByTheme: '/history/themes/:slug/:gameId',
  trackListing: '/history/games/:gameId',
  privacy: '/privacy',
});
