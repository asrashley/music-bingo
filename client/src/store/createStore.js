import { configureStore, getDefaultMiddleware } from '@reduxjs/toolkit';
import logger from 'redux-logger';
import { connectRouter } from 'connected-react-router';

import adminReducer from '../admin/adminSlice';
import directoriesReducer from '../directories/directoriesSlice';
import gamesReducer from '../games/gamesSlice';
import messagesReducer from '../messages/messagesSlice';
import settingsReducer from '../settings/settingsSlice';
import systemReducer from '../system/systemSlice';
import ticketsReducer from '../tickets/ticketsSlice';
import userReducer from '../user/userSlice';
import usersMiddleware from '../user/userMiddleware';
import messagesMiddleware from '../messages/messagesMiddleware';

import { history } from './history';

export function createStore(preloadedState, logging) {
  const middleware = [...getDefaultMiddleware(), usersMiddleware, messagesMiddleware];
  if (logging === undefined) {
    logging = (process.env.NODE_ENV === 'development');
  }
  if (logging) {
    middleware.push(logger);
  }
  return configureStore({
    middleware,
    reducer: {
      admin: adminReducer,
      directories: directoriesReducer,
      games: gamesReducer,
      messages: messagesReducer,
      settings: settingsReducer,
      system: systemReducer,
      tickets: ticketsReducer,
      router: connectRouter(history),
      user: userReducer,
    },
    preloadedState
  });
}
