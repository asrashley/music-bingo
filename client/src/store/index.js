export { createStore } from './createStore';
export { initialState } from './initialState';
export { history } from './history';
