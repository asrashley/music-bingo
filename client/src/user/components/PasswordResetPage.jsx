import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { reverse } from 'named-urls';


import { passwordResetUser } from '../userSlice';

import { getUser } from '../../user/userSelectors';

import routes from '../../routes';

import '../styles/user.scss';
import { PasswordResetForm } from './PasswordResetForm';
import { HistoryPropType } from '../../types/History';
import { UserPropType } from '../types/User';

export class PasswordResetPageComponent extends React.Component {
  static propTypes = {
    user: UserPropType.isRequired,
    dispatch: PropTypes.func,
    history: HistoryPropType.object,
  };

  constructor(props) {
    super(props);
    this.state = {
      resetSent: false,
      alert: '',
    };
  }

  componentDidMount() {
    const { user } = this.props;
    if (user.error) {
      this.setState({alert: user.error});
    }
  }

  componentDidUpdate(prevProps, prevState) {
    const { user } = this.props;
    if (user.error !== this.state.alert) {
      this.setState({alert: user.error});
    }
  }

  handleSubmit = ({ email }) => {
    const { dispatch, user } = this.props;
    if (user.isFetching === true) {
      return {
        type: "validate",
        message: "Password reset already is in progress",
        name: "email",
      };
    }
    if (user.pk > 0 && user.error === null) {
      return {
        type: "validate",
        message: "User is currently logged in",
        name: "email",
      };
    }
    return dispatch(passwordResetUser({ email }))
      .then((result) => {
        if(result && result.payload){
          const { payload } = result;
          if(payload.success === true){
            this.setState({ resetSent: true, email });
            return true;
          }
          this.setState({alert: payload.error});
          return false;
        }
        this.setState({ resetSent: true, email });
        return true;
      })
      .catch(err => {
        //console.error(err);
        if (err.error) {
          err = err.error;
        }
        this.setState({ alert: `${err}` });
        return {
          type: "validate",
          message: `${err}`,
          name: "email",
        };
      });
  };

  onCancel = () => {
    const { history } = this.props;
    history.push(reverse(`${routes.login}`));
  };

  render() {
    const { alert, email, resetSent } = this.state;
    const { user } = this.props;
    if (resetSent) {
      return (
        <div className="reset-sent">
          <h3>A password reset has been sent to {email}</h3>
          <p className="please-wait">Please wait for an email with instuctions on how to re-enable your account.</p>
        </div>
      );
    }
    return (
      <PasswordResetForm alert={alert} user={user} onSubmit={this.handleSubmit} onCancel={this.onCancel} />
    );
  }
};

const mapStateToProps = (state, ownProps) => {
  return {
    user: getUser(state, ownProps),
  };
};

export const PasswordResetPage = connect(mapStateToProps)(PasswordResetPageComponent);
