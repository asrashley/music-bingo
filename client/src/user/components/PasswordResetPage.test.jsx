import React from 'react';
import { screen, fireEvent } from '@testing-library/react';
import fetchMock from "fetch-mock-jest";
import log from 'loglevel';
import { reverse } from 'named-urls';

import routes from '../../routes';
import { createStore } from '../../store/createStore';
import { initialState } from '../../store/initialState';
import { renderWithProviders, installFetchMocks, setFormFields } from '../../testHelpers';
import { PasswordResetPage, PasswordResetPageComponent } from './PasswordResetPage';
import * as user from '../../fixtures/userState.json';

describe('PasswordResetPage component', () => {
  let apiMocks;

  beforeAll(() => {
    jest.useFakeTimers('modern');
    jest.setSystemTime(new Date('08 Feb 2023 10:12:00 GMT').getTime());
  });

  afterEach(() => {
    fetchMock.mockReset();
    log.resetLevel();
  });

  afterAll(() => jest.useRealTimers());

  beforeEach(() => {
    apiMocks = installFetchMocks(fetchMock, { loggedIn: false });
  });

  it('renders without throwing an exception with initial state', () => {
    const store = createStore(initialState);
    const props = {
      dispatch: store.dispatch,
      history: {
        push: jest.fn()
      },
      user
    };
    const result = renderWithProviders(<PasswordResetPage {...props} />);
    result.getByText("Request Password Reset");
    expect(result.asFragment()).toMatchSnapshot();
  });

  it('requests a password reset email', async () => {
    const store = createStore(initialState);
    const props = {
      dispatch: store.dispatch,
      history: {
        push: jest.fn()
      },
      user
    };
    const email = 'my.email@address.example';
    //log.setLevel('debug');
    const payload = await new Promise((resolve) => {
      fetchMock.post('/api/user/reset', (url, opts) => {
        const body = JSON.parse(opts.body);
        resolve(body);
        const { email } = body;
        return apiMocks.jsonResponse({
          email,
          success: true
        });
      });
      renderWithProviders(<PasswordResetPage {...props} />);
      setFormFields([{
        label: /^Email address/,
        value: email
      }]);
      fireEvent.click(screen.getByText("Request Password Reset"));
    });
    expect(payload).toEqual({
      email
    });
  });

  it('redirects if cancel is clicked', async () => {
    const url = await new Promise((resolve) => {
      const props = {
        user: initialState.user,
        dispatch: jest.fn(),
        history: {
          push: resolve
        }
      };
      renderWithProviders(<PasswordResetPage {...props} />);
      fireEvent.click(screen.getByText("Cancel"));
    });
    expect(url).toEqual(reverse(`${routes.login}`));
  });
});
