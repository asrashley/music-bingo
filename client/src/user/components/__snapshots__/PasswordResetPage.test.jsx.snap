// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordResetPage component renders without throwing an exception with initial state 1`] = `
<DocumentFragment>
  <form
    class="password-reset-form"
  >
    <h2>
      Request a password reset
    </h2>
    <div
      class="form-group"
    >
      <label
        for="field-email"
      >
        Email address
      </label>
      <input
        class="form-control   "
        id="field-email"
        name="email"
        placeholder="Email address"
        type="email"
      />
      <small
        class="form-text text-muted"
      >
        This must be the email address you used when registering. This is why we asked for your email address when you registered!
      </small>
    </div>
    <div
      class="form-group modal-footer"
    >
      <button
        class="btn btn-primary"
        type="submit"
      >
        Request Password Reset
      </button>
      <button
        class="btn btn-danger"
        name="cancel"
        type="cancel"
      >
        Cancel
      </button>
    </div>
  </form>
</DocumentFragment>
`;
